marketplace-service/src/main/java/com/axonivy/market/util/MavenUtils.java

```diff
@@ -0,0 +1,296 @@
+    String artifactName = artifact.getName();
+    if (StringUtils.isBlank(artifactName)) {
+      artifactName = GitHubUtils.convertArtifactIdToName(artifact.getArtifactId());
+    }
+    String type = StringUtils.defaultIfBlank(artifact.getType(), ProductJsonConstants.DEFAULT_PRODUCT_TYPE);
+    artifactName = String.format(MavenConstants.ARTIFACT_NAME_FORMAT, artifactName, type);
+    return Metadata.builder().groupId(artifact.getGroupId()).versions(new HashSet<>()).productId(productId).artifactId(
+        artifact.getArtifactId()).url(metadataUrl).repoUrl(
+        StringUtils.defaultIfEmpty(artifact.getRepoUrl(), MavenConstants.DEFAULT_IVY_MAVEN_BASE_URL)).type(type).name(
+        artifactName).isProductArtifact(BooleanUtils.isTrue(artifact.getIsProductArtifact())).build();
+  }
+
+  public static Metadata buildSnapShotMetadataFromVersion(Metadata metadata, String version) {
+    String snapshotMetadataUrl = buildSnapshotMetadataUrlFromArtifactInfo(metadata.getRepoUrl(), metadata.getGroupId(),
+        metadata.getArtifactId(), version);
+    return Metadata.builder().url(snapshotMetadataUrl).repoUrl(metadata.getRepoUrl()).groupId(
+        metadata.getGroupId()).artifactId(metadata.getArtifactId()).type(metadata.getType()).productId(
+        metadata.getProductId()).name(metadata.getName()).isProductArtifact(metadata.isProductArtifact()).build();
+  }
+
+  public static MavenArtifactModel buildMavenArtifactModelFromMetadata(String version, Metadata metadata) {
+    return new MavenArtifactModel(metadata.getName(),
+        buildDownloadUrl(metadata.getArtifactId(), version, metadata.getType(),
+            metadata.getRepoUrl(), metadata.getGroupId(), metadata.getSnapshotVersionValue()),
+        metadata.getArtifactId().contains(metadata.getGroupId()));
+  }
+
+  public static String getMetadataContentFromUrl(String metadataUrl) {
+    try {
+      return restTemplate.getForObject(metadataUrl, String.class);
+    } catch (Exception e) {
+      log.error("**MetadataService: Failed to fetch metadata from url {}", metadataUrl);
+      return StringUtils.EMPTY;
+    }
+  }
+
+  public static boolean isProductArtifactId (String artifactId) {
+    return StringUtils.endsWith(artifactId, MavenConstants.PRODUCT_ARTIFACT_POSTFIX);
+  }
+
+  public static Set<Metadata> convertArtifactsToMetadataSet(Set<Artifact> artifacts, String productId) {
+    Set<Metadata> results = new HashSet<>();
+    if (!CollectionUtils.isEmpty(artifacts)) {
+      artifacts.forEach(artifact -> {
+        String metadataUrl = buildMetadataUrlFromArtifactInfo(artifact.getRepoUrl(), artifact.getGroupId(),
+            artifact.getArtifactId());
+        results.add(convertArtifactToMetadata(productId, artifact, metadataUrl));
+        extractMetaDataFromArchivedArtifacts(productId, artifact, results);
+      });
+    }
+    return results;
+  }
+
+  public static void extractMetaDataFromArchivedArtifacts(String productId, Artifact artifact,
```

Comment:
please change return void to Metadata, then it makes the block code:         results.add(convertArtifactToMetadata(productId, artifact, metadataUrl));
extractMetaDataFromArchivedArtifacts(productId, artifact, results);
to 
results.add(convertArtifactToMetadata(productId, artifact, metadataUrl));
results.add(extractMetaDataFromArchivedArtifacts(productId, artifact));
easy to understand